### PR TITLE
[FormLayout] Added custom widths

### PR DESCRIPTION
**Pull request by ugur-shopify**
### WHY are these changes introduced?

Addresses https://github.com/Shopify/polaris-react/issues/4639

Our UX team is revising the form control widths in many parts of the Admin, and they are introducing a new paradigm where the form controls can have variable sizes (33%, 50%, 66%, 100%).

**Example 1**

**Example 2**

### WHAT is this pull request doing?

This PR extends the `FormLayout` component and allows the consumer to now specify `FormLayout.Item` width.

## 

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris-react/blob/master/README.md#development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Versioning%20and%20changelog.md)

Providing the Playground example below using the `Frame` component in order to be able to better demonstrate the container-responsive properties of the new `FormLayout` as the side nav expands / collapses.

<details>
<summary>Copy-paste this code in <code>playground/Playground.tsx</code>:</summary>

```jsx
import {
  ArrowLeftMinor,
  HomeMajor,
  OrdersMajor,
  ConversationMinor,
} from '@shopify/polaris-icons';
import React, {useCallback, useRef, useState} from 'react';

import {
  ActionList,
  AppProvider,
  Card,
  ContextualSaveBar,
  FormLayout,
  Frame,
  Layout,
  Loading,
  Modal,
  Navigation,
  SkeletonBodyText,
  SkeletonDisplayText,
  SkeletonPage,
  TextContainer,
  TextField,
  Toast,
  TopBar,
} from '../src';

const forms = (
  <>
    <Card title="One input">
      <Card.Section>
        <FormLayout>
          <FormLayout.Item size="small">
            <TextField
              type="text"
              label="Small"
              onChange={() => {}}
              autoComplete="off"
            />
          </FormLayout.Item>

          <FormLayout.Item size="medium">
            <TextField
              type="text"
              label="Medium"
              onChange={() => {}}
              autoComplete="off"
            />
          </FormLayout.Item>

          <FormLayout.Item size="large">
            <TextField
              type="text"
              label="Large"
              onChange={() => {}}
              autoComplete="off"
            />
          </FormLayout.Item>

          <FormLayout.Item size="full">
            <TextField
              type="text"
              label="Full"
              onChange={() => {}}
              autoComplete="off"
            />
          </FormLayout.Item>
        </FormLayout>
      </Card.Section>
    </Card>

    <Card title="Multiple Inputs">
      <Card.Section>
        <FormLayout>
          <FormLayout.Group>
            <FormLayout.Item size="small">
              <TextField
                type="text"
                label="Small"
                onChange={() => {}}
                autoComplete="off"
              />
            </FormLayout.Item>
            <FormLayout.Item size="small">
              <TextField
                type="text"
                label="Small"
                onChange={() => {}}
                autoComplete="off"
              />
            </FormLayout.Item>
          </FormLayout.Group>

          <FormLayout.Group>
            <FormLayout.Item size="medium">
              <TextField
                type="text"
                label="Medium"
                onChange={() => {}}
                autoComplete="off"
              />
            </FormLayout.Item>
            <FormLayout.Item size="small">
              <TextField
                type="text"
                label="Small"
                onChange={() => {}}
                autoComplete="off"
              />
            </FormLayout.Item>
          </FormLayout.Group>

          <FormLayout.Group>
            <FormLayout.Item size="large">
              <TextField
                type="text"
                label="Large"
                onChange={() => {}}
                autoComplete="off"
              />
            </FormLayout.Item>
            <FormLayout.Item size="small">
              <TextField
                type="text"
                label="Small"
                onChange={() => {}}
                autoComplete="off"
              />
            </FormLayout.Item>
          </FormLayout.Group>

          <FormLayout.Group>
            <FormLayout.Item size="small">
              <TextField
                type="text"
                label="Small"
                onChange={() => {}}
                autoComplete="off"
              />
            </FormLayout.Item>
            <FormLayout.Item size="large">
              <TextField
                type="text"
                label="Large"
                onChange={() => {}}
                autoComplete="off"
              />
            </FormLayout.Item>
          </FormLayout.Group>

          <FormLayout.Group>
            <FormLayout.Item size="small">
              <TextField
                type="text"
                label="Small"
                onChange={() => {}}
                autoComplete="off"
              />
            </FormLayout.Item>
            <FormLayout.Item size="small">
              <TextField
                type="text"
                label="Small"
                onChange={() => {}}
                autoComplete="off"
              />
            </FormLayout.Item>
            <FormLayout.Item size="small">
              <TextField
                type="text"
                label="Small"
                onChange={() => {}}
                autoComplete="off"
              />
            </FormLayout.Item>
          </FormLayout.Group>
        </FormLayout>
      </Card.Section>
    </Card>

    <Card title="Legacy / No size provided">
      <Card.Section title="One input">
        <TextField
          type="text"
          label="Legacy, full by default"
          onChange={() => {}}
          autoComplete="off"
        />
      </Card.Section>

      <Card.Section title="Grouped">
        <FormLayout>
          <FormLayout.Group>
            <TextField
              type="text"
              label="Legacy, full by default"
              onChange={() => {}}
              autoComplete="off"
            />

            <TextField
              type="text"
              label="Legacy, full by default"
              onChange={() => {}}
              autoComplete="off"
            />
          </FormLayout.Group>

          <FormLayout.Group>
            <TextField
              type="text"
              label="Legacy, full by default"
              onChange={() => {}}
              autoComplete="off"
            />

            <TextField
              type="text"
              label="Legacy, full by default"
              onChange={() => {}}
              autoComplete="off"
            />

            <TextField
              type="text"
              label="Legacy, full by default"
              onChange={() => {}}
              autoComplete="off"
            />
          </FormLayout.Group>
        </FormLayout>
      </Card.Section>

      <Card.Section title="Condensed">
        <FormLayout>
          <FormLayout.Group condensed>
            <TextField
              type="text"
              label="Legacy, full by default"
              onChange={() => {}}
              autoComplete="off"
            />

            <TextField
              type="text"
              label="Legacy, full by default"
              onChange={() => {}}
              autoComplete="off"
            />

            <TextField
              type="text"
              label="Legacy, full by default"
              onChange={() => {}}
              autoComplete="off"
            />

            <TextField
              type="text"
              label="Legacy, full by default"
              onChange={() => {}}
              autoComplete="off"
            />
          </FormLayout.Group>
        </FormLayout>
      </Card.Section>
    </Card>
  </>
);

export function Playground() {
  const defaultState = useRef({
    emailFieldValue: 'dharma@jadedpixel.com',
    nameFieldValue: 'Jaded Pixel',
  });
  const skipToContentRef = useRef(null);

  const [toastActive, setToastActive] = useState(false);
  const [isLoading, setIsLoading] = useState(false);
  const [isDirty, setIsDirty] = useState(false);
  const [searchActive, setSearchActive] = useState(false);
  const [searchValue, setSearchValue] = useState('');
  const [userMenuActive, setUserMenuActive] = useState(false);
  const [mobileNavigationActive, setMobileNavigationActive] = useState(false);
  const [modalActive, setModalActive] = useState(false);
  const [nameFieldValue, setNameFieldValue] = useState(
    defaultState.current.nameFieldValue,
  );
  const [emailFieldValue, setEmailFieldValue] = useState(
    defaultState.current.emailFieldValue,
  );
  const [storeName, setStoreName] = useState(
    defaultState.current.nameFieldValue,
  );
  const [supportSubject, setSupportSubject] = useState('');
  const [supportMessage, setSupportMessage] = useState('');

  const handleSubjectChange = useCallback(
    (value) => setSupportSubject(value),
    [],
  );
  const handleMessageChange = useCallback(
    (value) => setSupportMessage(value),
    [],
  );
  const handleDiscard = useCallback(() => {
    setEmailFieldValue(defaultState.current.emailFieldValue);
    setNameFieldValue(defaultState.current.nameFieldValue);
    setIsDirty(false);
  }, []);
  const handleSave = useCallback(() => {
    defaultState.current.nameFieldValue = nameFieldValue;
    defaultState.current.emailFieldValue = emailFieldValue;

    setIsDirty(false);
    setToastActive(true);
    setStoreName(defaultState.current.nameFieldValue);
  }, [emailFieldValue, nameFieldValue]);
  const handleNameFieldChange = useCallback((value) => {
    setNameFieldValue(value);
    value && setIsDirty(true);
  }, []);
  const handleEmailFieldChange = useCallback((value) => {
    setEmailFieldValue(value);
    value && setIsDirty(true);
  }, []);
  const handleSearchResultsDismiss = useCallback(() => {
    setSearchActive(false);
    setSearchValue('');
  }, []);
  const handleSearchFieldChange = useCallback((value) => {
    setSearchValue(value);
    setSearchActive(value.length > 0);
  }, []);
  const toggleToastActive = useCallback(
    () => setToastActive((toastActive) => !toastActive),
    [],
  );
  const toggleUserMenuActive = useCallback(
    () => setUserMenuActive((userMenuActive) => !userMenuActive),
    [],
  );
  const toggleMobileNavigationActive = useCallback(
    () =>
      setMobileNavigationActive(
        (mobileNavigationActive) => !mobileNavigationActive,
      ),
    [],
  );
  const toggleIsLoading = useCallback(
    () => setIsLoading((isLoading) => !isLoading),
    [],
  );
  const toggleModalActive = useCallback(
    () => setModalActive((modalActive) => !modalActive),
    [],
  );

  const toastMarkup = toastActive ? (
    <Toast onDismiss={toggleToastActive} content="Changes saved" />
  ) : null;

  const userMenuActions = [
    {
      items: [{content: 'Community forums'}],
    },
  ];

  const contextualSaveBarMarkup = isDirty ? (
    <ContextualSaveBar
      message="Unsaved changes"
      saveAction={{
        onAction: handleSave,
      }}
      discardAction={{
        onAction: handleDiscard,
      }}
    />
  ) : null;

  const userMenuMarkup = (
    <TopBar.UserMenu
      actions={userMenuActions}
      name="Dharma"
      detail={storeName}
      initials="D"
      open={userMenuActive}
      onToggle={toggleUserMenuActive}
    />
  );

  const searchResultsMarkup = (
    <ActionList
      items={[{content: 'Shopify help center'}, {content: 'Community forums'}]}
    />
  );

  const searchFieldMarkup = (
    <TopBar.SearchField
      onChange={handleSearchFieldChange}
      value={searchValue}
      placeholder="Search"
    />
  );

  const topBarMarkup = (
    <TopBar
      showNavigationToggle
      userMenu={userMenuMarkup}
      searchResultsVisible={searchActive}
      searchField={searchFieldMarkup}
      searchResults={searchResultsMarkup}
      onSearchResultsDismiss={handleSearchResultsDismiss}
      onNavigationToggle={toggleMobileNavigationActive}
    />
  );

  const navigationMarkup = (
    <Navigation location="/">
      <Navigation.Section
        items={[
          {
            label: 'Back to Shopify',
            icon: ArrowLeftMinor,
          },
        ]}
      />
      <Navigation.Section
        separator
        title="Jaded Pixel App"
        items={[
          {
            label: 'Dashboard',
            icon: HomeMajor,
            onClick: toggleIsLoading,
          },
          {
            label: 'Jaded Pixel Orders',
            icon: OrdersMajor,
            onClick: toggleIsLoading,
          },
        ]}
        action={{
          icon: ConversationMinor,
          accessibilityLabel: 'Contact support',
          onClick: toggleModalActive,
        }}
      />
    </Navigation>
  );

  const loadingMarkup = isLoading ? <Loading /> : null;

  const skipToContentTarget = (
    <a id="SkipToContentTarget" ref={skipToContentRef} tabIndex={-1} />
  );

  const actualPageMarkup = (
    <div style={{padding: '20px 4px'}}>
      <Layout>
        <Layout.Section>
          {skipToContentTarget}
          {forms}
        </Layout.Section>
      </Layout>
    </div>
  );

  const loadingPageMarkup = (
    <SkeletonPage>
      <Layout>
        <Layout.Section>
          <Card sectioned>
            <TextContainer>
              <SkeletonDisplayText size="small" />
              <SkeletonBodyText lines={9} />
            </TextContainer>
          </Card>
        </Layout.Section>
      </Layout>
    </SkeletonPage>
  );

  const pageMarkup = isLoading ? loadingPageMarkup : actualPageMarkup;

  const modalMarkup = (
    <Modal
      open={modalActive}
      onClose={toggleModalActive}
      title="Contact support"
      primaryAction={{
        content: 'Send',
        onAction: toggleModalActive,
      }}
    >
      <Modal.Section>
        <FormLayout>
          <TextField
            label="Subject"
            value={supportSubject}
            onChange={handleSubjectChange}
            autoComplete="off"
          />
          <TextField
            label="Message"
            value={supportMessage}
            onChange={handleMessageChange}
            autoComplete="off"
            multiline
          />
        </FormLayout>
      </Modal.Section>
    </Modal>
  );

  const theme = {
    logo: {
      width: 124,
      topBarSource:
        'https://cdn.shopify.com/s/files/1/0446/6937/files/jaded-pixel-logo-color.svg?6215648040070010999',
      contextualSaveBarSource:
        'https://cdn.shopify.com/s/files/1/0446/6937/files/jaded-pixel-logo-gray.svg?6215648040070010999',
      url: 'http://jadedpixel.com',
      accessibilityLabel: 'Jaded Pixel',
    },
  };

  return (
    <div style={{height: '500px'}}>
      <AppProvider
        theme={theme}
        i18n={{
          Polaris: {
            Avatar: {
              label: 'Avatar',
              labelWithInitials: 'Avatar with initials {initials}',
            },
            ContextualSaveBar: {
              save: 'Save',
              discard: 'Discard',
            },
            TextField: {
              characterCount: '{count} characters',
            },
            TopBar: {
              toggleMenuLabel: 'Toggle menu',

              SearchField: {
                clearButtonLabel: 'Clear',
                search: 'Search',
              },
            },
            Modal: {
              iFrameTitle: 'body markup',
            },
            Frame: {
              skipToContent: 'Skip to content',
              navigationLabel: 'Navigation',
              Navigation: {
                closeMobileNavigationLabel: 'Close navigation',
              },
            },
          },
        }}
      >
        <Frame
          topBar={topBarMarkup}
          navigation={navigationMarkup}
          showMobileNavigation={mobileNavigationActive}
          onNavigationDismiss={toggleMobileNavigationActive}
          // skipToContentTarget={skipToContentRef.current}
        >
          {contextualSaveBarMarkup}
          {loadingMarkup}
          {pageMarkup}
          {toastMarkup}
          {modalMarkup}
        </Frame>
      </AppProvider>
    </div>
  );
}
```

</details>

### 🎩 checklist

- [x] Tested on [mobile](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md#cross-browser-testing)
- [x] Tested on [multiple browsers](https://help.shopify.com/en/manual/intro-to-shopify/shopify-admin/supported-browsers)
- [ ] Tested for [accessibility](https://github.com/Shopify/polaris-react/blob/master/documentation/Accessibility%20testing.md)
- [x] Updated the component's `README.md` with documentation changes
- [x] [Tophatted documentation](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting%20documentation.md) changes in the style guide
- [ ] For visual design changes, ping @ sarahill to update the Polaris UI kit

### CODE REVIEW

File: src/components/FormLayout/FormLayout.scss [hunk 1]
```diff
@@ -36,4 +41,23 @@ $item-min-size: rem(220px);
     flex-basis: (0.5 * $item-min-size);
     min-width: (0.5 * $item-min-size);
   }
+
+  &.small {
+    flex-basis: #{$modifier-value};
+    max-width: Max(calc(33.33% - #{spacing(loose)}), #{$modifier-value});
```
Comment (ugur-shopify) [inline src/components/FormLayout/FormLayout.scss:47]: Using the standard Sass `max` throws an error `...is not a number for max` here. So, the `Max` solution is taken from [When Sass and New CSS Features Collide](https://css-tricks.com/when-sass-and-new-css-features-collide/#the-solution). Please let me know if there is an alternative to this and I can update.

File: src/components/FormLayout/FormLayout.scss [hunk 2]
```diff
@@ -1,6 +1,12 @@
+/* stylelint-disable function-name-case */
 @import '../../styles/common';
 
-$item-min-size: rem(220px);
+$item-min-size: rem(200px);
```
Comment (ugur-shopify) [inline src/components/FormLayout/FormLayout.scss:4]: Is changing the `$item-min-size` globally here really OK? I guess the alternative would be to change it only for the new `small`, `medium`, and `large` so it doesn't effect the pre-existing forms?

File: src/components/FormLayout/FormLayout.scss [hunk 3]
```diff
@@ -1,6 +1,12 @@
...
+$container-width: rem(684px);
+// This is a trick for container aware responsive design
+// Inspired from: https://heydonworks.com/article/the-flexbox-holy-albatross/
+$modifier: calc(#{$container-width} - 100%);
+$modifier-value: calc(#{$modifier} * 999);
```
Comment (ugur-shopify) [inline src/components/FormLayout/FormLayout.scss:9]: As outlined by the inline comments, this is a homebrew version of the `@container` queries. Not sure if there was any alternatives to this in our toolbox, and not sure if it's too hacky. It's been partially brought from [this blog post](https://heydonworks.com/article/the-flexbox-holy-albatross/).

Meanwhile, it is tested in BrowserStack using Chrome, Firefox, Safari, Opera, Edge, iOS Safari + Chrome, and Android Chrome on various devices, and they all work as expected.

**Comment by alex-page** [comment]
Hey @ugur-shopify, thanks for creating this PR. It's great to see how you want this functionality to work. This is a really ingenious solution. A few things top of my mind:
- The `FormLayout` component is something our team will remove with layout components in the future
- Lots of components in polaris have `small` `medium` `large` props and it is becoming an anti-pattern with missing layout components

If you want to proceed with this approach. Please know that we are looking into building a lot core functionality around layout.

**Comment by ugur-shopify** [comment]
@alex-page Thank you for your feedback!

Wondering what our next steps should be in this case. For context, we need this functionality in web as soon as possible.

Is it OK to move forward with the changes proposed here even though `FormLayout` component will eventually be retired? And by the way, what would be the alternative to `small`, `medium`, `large`?

cc @torben-nielsen

**Comment by alex-page** [comment]
I think this is fine to ship as is. I just want to make sure you are aware of the changes. There currently isn't an alternative to small/med/large. It's an anti pattern that indicates missing layout components. We need new components to stop them being created.

**Comment by ugur-shopify** [comment]
Sounds great, thank you! Switching from draft to `Ready for review`.

**Pull request closed without merging**